kv/txn_coord_sender.go

```diff
@@ -409,266 +389,85 @@ func (tc *TxnCoordSender) sendOne(ctx context.Context, call proto.Call) {
+	// - the commit timestamp is MaxOffset behind startNS
+	// - MaxOffset ns were spent in this function
+	// when returning to the client. Below we choose the option
+	// that involves less waiting, which is likely the first one
+	// unless a transaction commits with an odd timestamp.
+	if tsNS := br.Txn.Timestamp.WallTime; startNS > tsNS {
+		startNS = tsNS
 	}
-	return nil
-}
-
-// sendBatch unrolls a batched command and sends each constituent
-// command in parallel.
-// TODO(tschottdorf): modify sendBatch so that it sends truly parallel requests
-// when outside of a Transaction. This can then be used to address the TODO in
-// (*TxnCoordSender).resolve().
-func (tc *TxnCoordSender) sendBatch(ctx context.Context, batchArgs *proto.BatchRequest, batchReply *proto.BatchResponse) {
-	// Prepare the calls by unrolling the batch. If the batchReply is
-	// pre-initialized with replies, use those; otherwise create replies
-	// as needed.
-	// TODO(spencer): send calls in parallel.
-	batchReply.Txn = batchArgs.Txn
-	for i := range batchArgs.Requests {
-		args := batchArgs.Requests[i].GetValue().(proto.Request)
-		if err := updateForBatch(args, batchArgs.RequestHeader); err != nil {
-			batchReply.Header().SetGoError(err)
-			return
-		}
-		call := proto.Call{Args: args}
-		// Create a reply from the method type and add to batch response.
-		if i >= len(batchReply.Responses) {
-			call.Reply = args.CreateReply()
-			batchReply.Add(call.Reply)
-		} else {
-			call.Reply = batchReply.Responses[i].GetValue().(proto.Response)
-		}
-		tc.sendOne(ctx, call)
-		// Amalgamate transaction updates and propagate first error, if applicable.
-		if batchReply.Txn != nil {
-			batchReply.Txn.Update(call.Reply.Header().Txn)
-		}
-		if call.Reply.Header().Error != nil {
-			batchReply.Error = call.Reply.Header().Error
-			return
-		}
+	sleepNS := tc.clock.MaxOffset() -
+		time.Duration(tc.clock.PhysicalNow()-startNS)
+	if tc.linearizable && sleepNS > 0 {
+		defer func() {
+			if log.V(1) {
+				log.Infof("%v: waiting %s on EndTransaction for linearizability", br.Txn.Short(), util.TruncateDuration(sleepNS, time.Millisecond))
+			}
+			time.Sleep(sleepNS)
+		}()
+	}
+	if br.Txn.Status != proto.PENDING {
+		tc.cleanupTxn(trace, *br.Txn)
 	}
+	return br, nil
 }
 
-// updateResponseTxn updates the response txn based on the response
-// timestamp and error. The timestamp may have changed upon
-// encountering a newer write or read. Both the timestamp and the
-// priority may change depending on error conditions.
-func (tc *TxnCoordSender) updateResponseTxn(argsHeader *proto.RequestHeader, replyHeader *proto.ResponseHeader) {
-	// Move txn timestamp forward to response timestamp if applicable.
-	if replyHeader.Txn.Timestamp.Less(replyHeader.Timestamp) {
-		replyHeader.Txn.Timestamp = replyHeader.Timestamp
+// maybeBeginTxn begins a new transaction if a txn has been specified
+// in the request but has a nil ID. The new transaction is initialized
```

Comment:
That's what the client sends with the first request in a transaction.
On Aug 27, 2015 19:48, "Tamir Duberstein" <notifications@github.com> wrote:

> In kv/txn_coord_sender.go
> <https://github.com/cockroachdb/cockroach/pull/2141#discussion_r38124495>:
>
> >  }
> >
> > -// updateResponseTxn updates the response txn based on the response
> > -// timestamp and error. The timestamp may have changed upon
> > -// encountering a newer write or read. Both the timestamp and the
> > -// priority may change depending on error conditions.
> > -func (tc *TxnCoordSender) updateResponseTxn(argsHeader *proto.RequestHeader, replyHeader *proto.ResponseHeader) {
> > -	// Move txn timestamp forward to response timestamp if applicable.
> > -	if replyHeader.Txn.Timestamp.Less(replyHeader.Timestamp) {
> > -		replyHeader.Txn.Timestamp = replyHeader.Timestamp
> > +// maybeBeginTxn begins a new transaction if a txn has been specified
> > +// in the request but has a nil ID. The new transaction is initialized
>
> if a txn has been specified in the request but has a nil ID
>
> when/why does this happen?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2141/files#r38124495>.
>